src/com/owncloud/android/providers/DocumentsStorageProvider.java

```diff
@@ -0,0 +1,208 @@
+            throws FileNotFoundException {
+
+        final long folderId = Long.parseLong(parentDocumentId);
+        updateCurrentStorageManagerIfNeeded(folderId);
+
+        final FileCursor result = new FileCursor(projection);
+
+        final OCFile browsedDir = mCurrentStorageManager.getFileById(folderId);
+        for (OCFile file : mCurrentStorageManager.getFolderContent(browsedDir))
+            result.addFile(file);
+
+        return result;
+    }
+
+    @Override
+    public ParcelFileDescriptor openDocument(String documentId, String mode, CancellationSignal cancellationSignal)
+            throws FileNotFoundException {
+        final long docId = Long.parseLong(documentId);
+        updateCurrentStorageManagerIfNeeded(docId);
+
+        OCFile file = mCurrentStorageManager.getFileById(docId);
+
+        if (!file.isDown()) {
+
+            Intent i = new Intent(getContext(), FileDownloader.class);
+            i.putExtra(FileDownloader.EXTRA_ACCOUNT, mCurrentStorageManager.getAccount());
+            i.putExtra(FileDownloader.EXTRA_FILE, file);
+            getContext().startService(i);
+
+            do {
+                if (!waitOrGetCancelled(cancellationSignal))
+                    return null;
+                file = mCurrentStorageManager.getFileById(docId);
+
+            } while (!file.isDown());
+        }
+
+        return ParcelFileDescriptor.open(
+                new File(file.getStoragePath()), ParcelFileDescriptor.MODE_READ_ONLY);
+    }
+
+    @Override
+    public boolean onCreate() {
+        return true;
+    }
+
+    @Override
+    public AssetFileDescriptor openDocumentThumbnail(String documentId, Point sizeHint, CancellationSignal signal) throws FileNotFoundException {
+        long docId = Long.parseLong(documentId);
+        updateCurrentStorageManagerIfNeeded(docId);
+
+        OCFile file = mCurrentStorageManager.getFileById(docId);
+
+        File realFile = new File(file.getStoragePath());
+
+        return new AssetFileDescriptor(
+                ParcelFileDescriptor.open(realFile, ParcelFileDescriptor.MODE_READ_ONLY),
+                0,
+                AssetFileDescriptor.UNKNOWN_LENGTH);
+    }
+
+    @Override
+    public Cursor querySearchDocuments(String rootId, String query, String[] projection) throws FileNotFoundException {
+        updateCurrentStorageManagerIfNeeded(rootId);
+
+        OCFile root = mCurrentStorageManager.getFileByPath("/");
+        FileCursor result = new FileCursor(projection);
+
+        for (OCFile f : findFiles(root, query))
+            result.addFile(f);
+
+        return result;
+    }
+
+    private void updateCurrentStorageManagerIfNeeded(long docId) {
+        if (mCurrentStorageManager == null ||
+                (mRootIdToStorageManager.containsKey(docId) &&
+                        mCurrentStorageManager != mRootIdToStorageManager.get(docId))) {
+            mCurrentStorageManager = mRootIdToStorageManager.get(docId);
```

Comment:
The map is always instantiated because of how DocumentsProvider is calling methods in this class. It is guaranteed that `queryRoots` will always be called before anything else and it is the place where the map is instantiated.